hist/hist/inc/THnChain.h

```diff
@@ -0,0 +1,266 @@
+   {
+      // Forwards to `THnBase::ProjectionND()`.
+      return static_cast<THnBase*>(ProjectionAny(ndim, dim, option));
+   }
+
+ private:
+   TString fName; ///< name of the histogram
+
+   TObjArray fFiles; ///< a list of files to extract the histogram from
+   TObjArray fAxes;  ///< the list of histogram axes
+
+   /// Projects all histograms in the chain.
+   ///
+   /// See `THnBase::Projection` for parameters and their semantics.
+   TObject* ProjectionAny(Int_t ndim, const Int_t* dim, Option_t* option = "")
+   {
+      if (ndim == 0) {
+         return nullptr;
+      }
+
+      TObject* h_merged = nullptr;
+      for (const auto& obj : fFiles) {
+         THnBase* hs = ReadHistogram(static_cast<TObjString*>(obj)->String());
+
+         if (!hs || !CheckConsistency(*hs, fAxes)) {
+            Warning("ProjectionAny",
+                    "Could find histogram matching template %s in file %s",
+                    fName.Data(),
+                    static_cast<TObjString*>(obj)->String().Data());
+            continue;
+         }
+
+         SetupAxes(*hs);
+
+         // Perform projection.
+         TObject* h = nullptr;
+
+         if (ndim == 1) {
+            h = static_cast<TObject*>(hs->Projection(dim[0], option));
+         } else if (ndim == 2) {
+            h = static_cast<TObject*>(hs->Projection(dim[0], dim[1], option));
+         } else if (ndim == 3) {
+            h = static_cast<TObject*>(hs->Projection(dim[0], dim[1], dim[2], option));
+         } else {
+            h = static_cast<TObject*>(hs->ProjectionND(ndim, dim, option));
+         }
+
+         delete hs;
+
+         // Add this histogram.
+         if (h_merged) {
+            if (ndim < 3) {
+               static_cast<TH1*>(h_merged)->Add(static_cast<TH1*>(h));
+            } else {
+               static_cast<THnBase*>(h_merged)->Add(static_cast<THnBase*>(h));
+            }
+
+            delete h;
+         } else {
+            h_merged = h;
+         }
+      }
+
+      return h_merged;
+   }
+
+   /// Retrieve a histogram from a file.
+   ///
+   /// \param fileName path of the file to read.
+   THnBase* ReadHistogram(const char* fileName)
+   {
+      TFile* f = TFile::Open(fileName);
+      f->cd();
+      THnBase* hs = nullptr;
+      f->GetObject(fName, hs);
+      delete f;
+
+      return hs;
+   }
+
+   /// Copy the properties of all axes a histogram.
+   ///
+   /// \param hs histogram whose axes should be updated
+   void SetupAxes(THnBase& hs)
+   {
+      const Int_t naxes = fAxes.GetEntries();
+      for (Int_t i = 0; i < naxes; ++i) {
+         const TAxis* ax_ref = static_cast<TAxis*>(fAxes[i]);
+         TAxis* ax = hs.GetAxis(i);
+         ax_ref->Copy(*ax);
+      }
+   }
+
+   /// Ensure a histogram has axes similar to the ones we expect.
+   ///
+   /// \param h histogram to verify
+   /// \param axes expected set of axes
+   static bool CheckConsistency(const THnBase& h, const TObjArray& axes)
```

Comment:
Can we move these into the cxx? I'd like to remove the amount of #includes one gets when including THnChain.h